src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,107 @@
+        }
+        values[sizeOfArray] = value;
+        sizeOfArray++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > sizeOfArray || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index
+                    + " is greater than array size: " + sizeOfArray);
+        }
+        System.arraycopy(values, index, values, index + 1, sizeOfArray - index);
+        values[index] = value;
+        sizeOfArray++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        T[] array = (T[]) new Object[list.size()];
+        for (int i = 0; i < list.size(); i++) {
+            array[i] = list.get(i);
+        }
+        System.arraycopy(array, 0, values, sizeOfArray, list.size());
+        sizeOfArray = sizeOfArray + array.length;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= sizeOfArray || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Element does not exist");
+        }
+        return values[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= sizeOfArray || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index
+                    + " is greater than array size: " + sizeOfArray);
+        }
+        values[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T[] oldArray = (T[]) new Object[sizeOfArray];
+        System.arraycopy(values, 0, oldArray, 0, sizeOfArray);
+        if (index >= sizeOfArray || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Element does not exist");
+        }
+        sizeOfArray--;
+        T oldValue = values[index];
+        System.arraycopy(oldArray, 0, values, 0, index);
+        System.arraycopy(oldArray, index + 1, values, index, oldArray.length - (index + 1));
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T[] oldArray = (T[]) new Object[sizeOfArray];
+        System.arraycopy(values, 0, oldArray, 0, sizeOfArray);
+        for (int i = 0; i < sizeOfArray; i++) {
+            if (values[i] == element || values[i] != null && values[i].equals(element)) {
+                sizeOfArray--;
+                T oldValue = values[i];
+                System.arraycopy(oldArray, 0, values, 0, i);
+                System.arraycopy(oldArray, i + 1, values, i, oldArray.length - (i + 1));
+                return oldValue;
+            }
+        }
+        throw new NoSuchElementException("Element with value " + element + " does not exist");
     }
 
     @Override
     public int size() {
-        return 0;
+        return sizeOfArray;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return sizeOfArray == 0;
+    }
+
+    private int grow(int size) {
+        return (int) (size * 1.5);
     }
```

Comment:
this logic is so small :)
let's refactor this method - maybe we can move a larger piece of code here